Typo3 12 compatibility

hojalatheef opened the pull request.

hojalatheef committed:
[TASK] Compatibility Fixes 12 Started

hojalatheef committed:
Compatibility fixes

hojalatheef committed:
Configurations update

hojalatheef committed:
[TASK] Fixed repository queries

hojalatheef committed:
Compatibility Fixes for TYPO3 12LTS

hojalatheef committed:
Scheduler compatibility fixes

hojalatheef committed:
[TASK] Compatibility Fixes for TYPO3 Version 12 LTS

- New git workflow
- TCA Migrations
- Scheduler Task Compatibility Fixes
- Deprecated Usages fixed
- Unwanted JS usages removed

hojalatheef committed:
Functional Test Case Fixes

hojalatheef committed:
gitignore updated

hojalatheef committed:
Removed XML Data Fixtures and replaced with CSV

hojalatheef committed:
Code formatting and php lint issues fixed

hojalatheef committed:
wrong extension name in github pipeline

hojalatheef committed:
Add testing docker configration file

hojalatheef committed:
CGL fix

hojalatheef committed:
CGL fix

hojalatheef committed:
CGL fix

hojalatheef committed:
BUGFIX for postgress db type

sfroemkenjw commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,53 @@
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
```
Checkout v4 is out now

hojalatheef replied:
Done ☑️

sfroemkenjw commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,53 @@
-        include:
-        - typo3: ^11.5
-          php: '8.0'
+        php: [ '8.1', '8.2' ]
```
You change my "pull_request" from line 4 into 2 lines, but here you collect the options as one-liner ;-)

sfroemkenjw commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,53 @@
-          git checkout composer.json
-      - name: php-cs-fixer
-        run: composer ci:php:fixer
+        uses: actions/checkout@v3
```
v4

sfroemkenjw commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,53 @@
+      - name: Composer
+        run: Build/Scripts/runTests.sh -p ${{ matrix.php }} -t 12 -s composerUpdate
 
       - name: Lint PHP
```
Remove linting section. Was already processed in CGL part

sfroemkenjw commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,53 @@
 
-      - name: Unit Tests
-        run: composer ci:tests:unit
+      - name: Functional tests with sqlite
```
I'm missing a call to test the existing UnitTests

sfroemkenjw commented on Build/Scripts/runTests.sh:
```diff
@@ -0,0 +1,395 @@
+
+# Load help text into $HELP
+read -r -d '' HELP <<EOF
+jweiland replacer test runner. Execute unit test suite and some other details.
```
I'm pretty sure: This is not the test runner for EXT:replacer ;-)

sfroemkenjw commented on Build/FunctionalTestsBootstrap.php:
```diff
@@ -0,0 +1,17 @@
+<?php
+
+/*
+ * This file is part of the package jweiland/replacer.
```
Wrong extkey

sfroemkenjw commented on Build/UnitTestsBootstrap.php:
```diff
@@ -0,0 +1,75 @@
+<?php
+
+/*
+ * This file is part of the package jweiland/replacer.
```
Wrong extkey

sfroemkenjw commented on Tests/Functional/Fixtures/DummyUserFunc.php:
```diff
@@ -0,0 +1,20 @@
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace JWeiland\Replacer\Tests\Functional\Fixtures;
```
Wrong extkey

sfroemkenjw commented on Tests/Functional/Fixtures/basic_template.typoscript:
```diff
@@ -0,0 +1,24 @@
+config.tx_replacer {
```
Configuration for replacer? I thought you want to upgrade weather2.

sfroemkenjw commented on Tests/Functional/Fixtures/user_int.typoscript:
```diff
@@ -0,0 +1,2 @@
+page.20 = USER_INT
+page.20.userFunc = JWeiland\Replacer\Tests\Functional\Fixtures\DummyUserFunc->render
```
Wrong namespace. Wrong extkey

sfroemkenjw commented on Build/php-cs-fixer/config.php:
```diff
@@ -20,12 +23,12 @@
     ->in(__DIR__);
 
-return (new \PhpCsFixer\Config())
+return (new Config())
```
Please add a test for correct class header. Example: https://github.com/froemken/mysqlreport/blob/main/Build/php-cs-fixer/config.php#L39-L41

hojalatheef replied:
It's already there in the test. 
Line : 37-39
`'header_comment' => [
            'header' => $headerComment,
        ],`

sfroemkenjw commented on Classes/Dashboard/Provider/WeatherAlertDataProvider.php:
```diff
@@ -32,9 +32,7 @@ private function getWeatherAlerts(): array
-            ->select('*')
-            ->from('tx_weather2_domain_model_weatheralert')
-            ->execute();
+            ->select('*')->from('tx_weather2_domain_model_weatheralert')->executeQuery();
```
Concatenate methods is OK, but not as never ending one-liners.

sfroemkenjw commented on Classes/Domain/Repository/DwdWarnCellRepository.php:
```diff
@@ -36,10 +36,7 @@ public function findByName(string $name): array
-                        $query->like('name', '%' . trim($name) . '%'),
-                        $query->equals('warn_cell_id', $name)
-                    )
+                    $query->logicalOr(...[$query->like('name', '%' . trim($name) . '%'), $query->equals('warn_cell_id', $name)])
```
Please wrap into multiple lines

sfroemkenjw commented on Classes/Domain/Repository/WeatherAlertRepository.php:
```diff
@@ -51,17 +52,14 @@ public function findByUserSelection(
+            $andConstraints[] = $query->logicalOr(...$warningConstraints['type']);
+            $andConstraints[] = $query->logicalOr(...$warningConstraints['level']);
+            $andConstraints[] = $query->logicalOr(...$warnCellConstraints);
+            $andConstraints[] = $query->logicalOr(...[$query->greaterThanOrEqual('end_date', GeneralUtility::makeInstance(Context::class)->getPropertyFromAspect('date', 'timestamp')), $query->equals('end_date', 0)]);
```
Please wrap into multiple lines

sfroemkenjw commented on Classes/Task/DeutscherWetterdienstTask.php:
```diff
@@ -35,11 +34,6 @@ class DeutscherWetterdienstTask extends AbstractTask
-    /**
-     * @var ObjectManager
-     */
-    protected $objectManager;
```
If you remove properties from a task, don't forget that these tasks still exist in a serialized version in tx_scheduler_task table WITH this old ObjectManager property. So please add a note to the upgrade wizard section of docs, that this kind of task has to be deleted and re-created.

sfroemkenjw commented on Classes/Task/DeutscherWetterdienstTask.php:
```diff
@@ -95,8 +89,7 @@ class DeutscherWetterdienstTask extends AbstractTask
     {
-        $this->objectManager = GeneralUtility::makeInstance(ObjectManager::class);
-        $this->dwdWarnCellRepository = $this->objectManager->get(DwdWarnCellRepository::class);
+        $this->dwdWarnCellRepository = GeneralUtility::makeInstance(DwdWarnCellRepository::class);
```
As this is a major upgrade I prefer to use this possibility to remove ALL properties of that class for better compatibility with future TYPO3 versions. Please create getter methods for all these needed classes.

sfroemkenjw commented on Classes/Task/DeutscherWetterdienstTask.php:
```diff
@@ -139,7 +132,7 @@ protected function decodeResponse(ResponseInterface $response): array
     protected function handleResponse(): void
     {
-        $this->persistenceManager = $this->objectManager->get(PersistenceManager::class);
+        $this->persistenceManager = GeneralUtility::makeInstance(PersistenceManager::class);
```
See comment above

sfroemkenjw commented on Classes/Task/DeutscherWetterdienstTaskAdditionalFieldProvider.php:
```diff
@@ -98,14 +97,14 @@ public function getAdditionalFields(
                 WeatherUtility::translate('message.noDwdWarnCellsFound', 'deutscherwetterdienst'),
                 '',
-                AbstractMessage::WARNING
+                \TYPO3\CMS\Core\Type\ContextualFeedbackSeverity::WARNING
```
Add a "use" definition and shorten the class name

sfroemkenjw commented on Classes/Task/DeutscherWetterdienstTaskAdditionalFieldProvider.php:
```diff
@@ -116,9 +115,9 @@ public function getAdditionalFields(
-size="30" placeholder="' . WeatherUtility::translate('placeholder.recordStoragePage', 'openweatherapi') . ' --->"/><span class="input-group-btn"><a href="#" class="btn btn-default" onclick="TYPO3.FormEngine.openPopupWindow(\'db\',\'tx_scheduler[dwd_recordStoragePage]|||pages|\'); return false;">' .
-            WeatherUtility::translate('buttons.recordStoragePage', 'deutscherwetterdienst') . '</a></span></div>';
+        $fieldCode = '<div class="input-group">
+            <input type="text" class="form-control" name="tx_scheduler[' . $fieldID . ']" id="' . $fieldID . '" value="' . $taskInfo[$fieldID] . '" size="30" placeholder="' . WeatherUtility::translate('placeholder.recordStoragePage', 'openweatherapi') . ' --->"/>
```
Not tested, but as you have removed the button from above there is no need to add this arrow "--->", right?

sfroemkenjw commented on Classes/Task/DeutscherWetterdienstTaskAdditionalFieldProvider.php:
```diff
@@ -142,6 +141,7 @@ protected function initialize(): void
         );
         $this->pageRenderer->addCssFile('EXT:weather2/Resources/Public/Css/dwdScheduler.css');
         $this->pageRenderer->loadRequireJsModule('TYPO3/CMS/Backend/FormEngineValidation');
+        $this->pageRenderer->loadRequireJsModule('TYPO3/CMS/Weather2/jquery.autocomplete');
```
You have remove ui-autocomplete-input class from DeutscherWetterdienstTaskAdditionalFieldProvider class. Are you sure this line is still needed?

hojalatheef replied:
No, Autocomplete for the alert is not removed. It's working and this is needed as jquery/autocomplete is removed from TYPO3 Core (12).

sfroemkenjw commented on Classes/Task/DeutscherWetterdienstTaskAdditionalFieldProvider.php:
```diff
@@ -207,7 +206,7 @@ public function validateAdditionalFields(array &$submittedData, SchedulerModuleC
             if (empty($value) && in_array($fieldName, $this->requiredFields, true)) {
                 $isValid = false;
-                $this->addMessage('Field: ' . $fieldName . ' must not be empty', AbstractMessage::ERROR);
+                $this->addMessage('Field: ' . $fieldName . ' must not be empty', \TYPO3\CMS\Core\Type\ContextualFeedbackSeverity::ERROR);
```
Add a "use" definition and shorten the namespace

sfroemkenjw commented on Classes/Task/OpenWeatherMapTask.php:
```diff
@@ -110,7 +110,12 @@ class OpenWeatherMapTask extends AbstractTask
+    /**
+     * @var PersistenceManager
+     */
+    public $persistenceManager = '';
```
I'm pretty sure the $persistenceManager does not defaults to an empty string

sfroemkenjw commented on Classes/Task/OpenWeatherMapTask.php:
```diff
@@ -155,11 +160,10 @@ public function execute(): bool
 
-        $persistenceManager->add($this->getCurrentWeatherInstanceForResponseClass($this->responseClass));
-        $persistenceManager->persistAll();
+        $this->persistenceManager = GeneralUtility::makeInstance(PersistenceManager::class);
```
As above: Try to create getters for these objects and remove all possible properties from this class.

sfroemkenjw commented on Classes/Task/OpenWeatherMapTask.php:
```diff
@@ -155,11 +160,10 @@ public function execute(): bool
-        $persistenceManager->persistAll();
+        $this->persistenceManager = GeneralUtility::makeInstance(PersistenceManager::class);
+        $this->persistenceManager->add($this->getCurrentWeatherInstanceForResponseClass($this->responseClass));
+        $this->persistenceManager->persistAll();
```
Using extbase in tasks where no TSFE resides is a very bad idea. Do you see an easy way to migrate that to QueryBuilder? If not...we should migrate that within next major release.

sfroemkenjw commented on Classes/Task/OpenWeatherMapTask.php:
```diff
@@ -345,4 +349,9 @@ protected function removeOldRecordsFromDb(): void
         );
     }
+
+    public function setPersistenceManager($persistenceManager)
```
"set"?!?!

hojalatheef replied:
I put that for some test. I forget to remove it.

sfroemkenjw commented on Classes/Task/OpenWeatherMapTaskAdditionalFieldProvider.php:
```diff
@@ -75,11 +77,12 @@ class OpenWeatherMapTaskAdditionalFieldProvider extends AbstractAdditionalFieldP
         PageRenderer $pageRenderer
     ) {
-        $this->countryRepository = $countryRepository;
+        //$this->countryRepository = $countryRepository;
```
Remove. We have git for history

sfroemkenjw commented on Classes/Task/OpenWeatherMapTaskAdditionalFieldProvider.php:
```diff
@@ -225,7 +226,7 @@ public function validateAdditionalFields(
             if (empty($value) && in_array($fieldName, $this->requiredFields, true)) {
                 $isValid = false;
-                $this->addMessage('Field: ' . $fieldName . ' can not be empty', AbstractMessage::ERROR);
+                $this->addMessage('Field: ' . $fieldName . ' can not be empty', \TYPO3\CMS\Core\Type\ContextualFeedbackSeverity::ERROR);
```
shorten namespace

sfroemkenjw commented on Classes/Task/OpenWeatherMapTaskAdditionalFieldProvider.php:
```diff
@@ -261,21 +262,21 @@ private function isValidResponseCode(
             $this->addMessage(
                 WeatherUtility::translate('message.api_response_401', 'openweatherapi'),
-                AbstractMessage::ERROR
+                \TYPO3\CMS\Core\Type\ContextualFeedbackSeverity::ERROR
```
short namespace

sfroemkenjw commented on Classes/Task/OpenWeatherMapTaskAdditionalFieldProvider.php:
```diff
@@ -261,21 +262,21 @@ private function isValidResponseCode(
             $this->addMessage(
                 WeatherUtility::translate('message.api_code_404', 'openweatherapi'),
-                AbstractMessage::ERROR
+                \TYPO3\CMS\Core\Type\ContextualFeedbackSeverity::ERROR
```
short namespace

sfroemkenjw commented on Classes/Task/OpenWeatherMapTaskAdditionalFieldProvider.php:
```diff
@@ -261,21 +262,21 @@ private function isValidResponseCode(
             $this->addMessage(
                 WeatherUtility::translate('message.api_response_null', 'openweatherapi'),
-                AbstractMessage::ERROR
+                \TYPO3\CMS\Core\Type\ContextualFeedbackSeverity::ERROR
```
short namespace

sfroemkenjw commented on Classes/Task/OpenWeatherMapTaskAdditionalFieldProvider.php:
```diff
@@ -292,14 +293,14 @@ private function isValidResponseCode(
                 $this->addMessage(
                     WeatherUtility::translate('message.api_code_404', 'openweatherapi'),
-                    AbstractMessage::ERROR
+                    \TYPO3\CMS\Core\Type\ContextualFeedbackSeverity::ERROR
```
short namespace

sfroemkenjw commented on Classes/Task/OpenWeatherMapTaskAdditionalFieldProvider.php:
```diff
@@ -292,14 +293,14 @@ private function isValidResponseCode(
                     WeatherUtility::translate('message.api_code_none', 'openweatherapi'),
                     json_encode($responseClass)
-                ), AbstractMessage::ERROR);
+                ), \TYPO3\CMS\Core\Type\ContextualFeedbackSeverity::ERROR);
```
short namespace

sfroemkenjw commented on Classes/Task/OpenWeatherMapTaskAdditionalFieldProvider.php:
```diff
@@ -338,18 +339,20 @@ private function checkMailConfiguration(): string
-        $countries = $this->countryRepository->findAll();
-
+        $countries = $this->countryProvider->getAll();
+        //debug($countries);die;
```
Remove debug entries

sfroemkenjw commented on Classes/Upgrade/EmptyTaskLoggerUpgrade.php:
```diff
@@ -112,28 +112,19 @@ protected function getWeather2SchedulerTasks(): array
-                )
-            )
-            ->execute();
+            ->from('tx_scheduler_task')->where($queryBuilder->expr()->or($queryBuilder->expr()->like(
```
wrap this into multiple lines

sfroemkenjw commented on Classes/Upgrade/EmptyTaskLoggerUpgrade.php:
```diff
@@ -112,28 +112,19 @@ protected function getWeather2SchedulerTasks(): array
+            ->from('tx_scheduler_task')->where($queryBuilder->expr()->or($queryBuilder->expr()->like(
+                'serialized_task_object',
+                $queryBuilder->createNamedParameter('%OpenWeatherMapTask%')
+            ), $queryBuilder->expr()->like(
```
Move each following expression into a new line

sfroemkenjw commented on Configuration/TCA/tx_weather2_domain_model_weatheralert.php:
```diff
@@ -77,20 +77,21 @@
                 ],
                 'size' => 1,
-                'eval' => 'int,required',
+                'eval' => 'int',
```
I think we can remove that as it is a selectbox with fixed values

sfroemkenjw commented on Configuration/TCA/tx_weather2_domain_model_weatheralert.php:
```diff
@@ -100,23 +101,24 @@
                 ],
                 'size' => 1,
-                'eval' => 'int,required',
+                'eval' => 'int',
```
Remove

sfroemkenjw commented on Documentation/AdministratorManual/Upgrade/Index.rst:
```diff
@@ -7,6 +7,14 @@ Upgrade
+Add TYPO3 12 Compatibility
+Removed older version compatibility
+Removed dependency with EXT:static_info_tables
+
```
Add a note about re-creating the Scheduler Tasks

sfroemkenjw commented on Resources/Public/JavaScript/DeutscherWetterdienstTaskModule.js:
```diff
@@ -2,7 +2,7 @@
  * Object that replace pages_ID with ID
  */
-define('TYPO3/CMS/Weather2/DeutscherWetterdienstTaskModule', ['jquery', 'jquery/autocomplete'], function ($) {
+define('TYPO3/CMS/Weather2/DeutscherWetterdienstTaskModule', ['jquery'], function ($) {
```
As the button and AutoComplete feature is gone now, do we still need that file?

hojalatheef replied:
Autocomplete is still working :) That's why I loaded the autocomplete from inside the plugin.

sfroemkenjw commented on Resources/Public/JavaScript/jquery.autocomplete.js:
```diff
@@ -5,7 +5,6 @@
  *  For details, see the web site: https://github.com/devbridge/jQuery-Autocomplete
  */
-
 // Expose plugin as an AMD module if AMD loader is present:
```
As the button and AutoComplete feature is gone now, do we still need that file?

sfroemkenjw commented on Tests/Functional/Domain/Repository/CurrentWeatherRepositoryTest.php:
```diff
@@ -26,30 +25,13 @@ class CurrentWeatherRepositoryTest extends FunctionalTestCase
-        'typo3conf/ext/weather2',
-        'typo3conf/ext/static_info_tables',
-    ];
+    protected array $testExtensionsToLoad = ['typo3conf/ext/weather2'];
```
Since testingframework 7.0.3 is has to the package name. And please wrap it into multiple lines.

sfroemkenjw commented on Tests/Functional/Domain/Repository/DwdWarnCellRepositoryTest.php:
```diff
@@ -25,19 +25,13 @@ class DwdWarnCellRepositoryTest extends FunctionalTestCase
-        'typo3conf/ext/weather2',
-        'typo3conf/ext/static_info_tables',
-    ];
+    protected array $testExtensionsToLoad = ['typo3conf/ext/weather2'];
```
package name

sfroemkenjw commented on Tests/Functional/Domain/Repository/WeatherAlertRepositoryTest.php:
```diff
@@ -26,19 +26,13 @@ class WeatherAlertRepositoryTest extends FunctionalTestCase
-        'typo3conf/ext/weather2',
-        'typo3conf/ext/static_info_tables',
-    ];
+    protected array $testExtensionsToLoad = ['typo3conf/ext/weather2'];
```
package name

sfroemkenjw commented on Tests/Functional/Task/OpenWeatherMapTaskTest.php:
```diff
@@ -46,37 +43,21 @@ class OpenWeatherMapTaskTest extends FunctionalTestCase
-    protected $coreExtensionsToLoad = [
-        'scheduler',
-    ];
+    protected array $coreExtensionsToLoad = ['scheduler'];
```
multi line

sfroemkenjw commented on Tests/Functional/Task/OpenWeatherMapTaskTest.php:
```diff
@@ -46,37 +43,21 @@ class OpenWeatherMapTaskTest extends FunctionalTestCase
-        'typo3conf/ext/weather2',
-        'typo3conf/ext/static_info_tables',
-    ];
+    protected array $testExtensionsToLoad = ['typo3conf/ext/weather2'];
```
Use package name and multi line

sfroemkenjw commented on composer.json:
```diff
@@ -18,13 +18,16 @@
-		"typo3/cms-core": "^10.4.29 || ^11.5.11",
-		"sjbr/static-info-tables": ">=6.6.0"
+		"typo3/cms-core": "^12.4",
+		"typo3/cms-dashboard": "^12.4",
```
I can't find a dashboard widget in Services.yaml. Why EXT:dashboard is needed?

hojalatheef replied:
Its imported at the top of the file using 
`imports:
  { resource: Backend/DashboardWidgets.yaml }`

hojalatheef committed:
[BUGFIX] GITHUB Pipeline Fixes

- updated action version to 4
- added Unit Tests to testsuite

hojalatheef committed:
[BUGFIX] Build/corrected incorrect package name

hojalatheef committed:
[BUGFIX] Scheduler Task Fixes

- code review corrections made
- seperate getter for classes write down in a seperate WeatherAbstract Class
- wrapping corrections
- shorten class names with use

hojalatheef committed:
[BUGFIX] Code formatting issues

hojalatheef committed:
[BUGFIX] Code Formatting Issues

hojalatheef committed:
Documentation Upgrade Info added

hojalatheef committed:
[BUGFIX] package name changed to vendor/package format

hojalatheef committed:
[BUGFIX] Making the WeatherAbstract class to abstract

hojalatheef committed:
[TASK] Removed Upgrade Wizard, OpenWeatherMapTask insertion changed to ConnectionPool query

- removed upgrade wizard
- insert weather object moved from persistenceManager to ConnectionPool Query
- test case added for asserting the insertion with connection pool query

hojalatheef committed:
TestCase fix

hojalatheef committed:
Fix with test case
